crates/interfaces/src/p2p/full_block.rs

```diff
@@ -192,28 +235,307 @@ where
+                .zip(bodies.iter())
+                .map(|(h, b)| SealedBlock::new(h.clone(), b.clone()))
+                .collect::<Vec<_>>(),
+        )
+    }
+
+    /// Returns whether or not a bodies request has been started, returning false if there is no
+    /// pending request, and if there is no buffered response.
+    fn has_bodies_request_started(&self) -> bool {
+        self.request.bodies.is_some() && self.bodies.is_some()
+    }
+}
+
+impl<Client> Future for FetchFullBlockRangeFuture<Client>
+where
+    Client: BodiesClient + HeadersClient + Unpin + 'static,
+{
+    type Output = Vec<SealedBlock>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        loop {
+            match ready!(this.request.poll(cx)) {
+                // This branch handles headers responses from peers - it first ensures that the
+                // starting hash and number of headers matches what we requested.
+                //
+                // If these don't match, we penalize the peer and retry the request.
+                // If they do match, we sort the headers by block number and start the request for
+                // the corresponding block bodies.
+                //
+                // The next result that should be yielded by `poll` is the bodies response.
+                RangeResponseResult::Header(res) => {
+                    match res {
+                        Ok(headers) => {
+                            let (peer, mut headers) = headers
+                                .map(|h| {
+                                    h.iter().map(|h| h.clone().seal_slow()).collect::<Vec<_>>()
+                                })
+                                .split();
+
+                            // ensure the response is what we requested
+                            if headers.is_empty() || (headers.len() as u64) != this.count {
+                                // received bad response
+                                this.client.report_bad_message(peer);
+                            } else {
+                                // sort headers from highest to lowest block number
+                                headers.sort_unstable_by_key(|h| Reverse(h.number));
+
+                                // check the starting hash
+                                if headers[0].hash() != this.hash {
+                                    // received bad response
+                                    this.client.report_bad_message(peer);
+                                } else {
+                                    // get the bodies request so it can be polled later
+                                    let hashes =
+                                        headers.iter().map(|h| h.hash()).collect::<Vec<_>>();
+
+                                    // set the actual request if it hasn't been started yet
+                                    if !this.has_bodies_request_started() {
```

Comment:
is this even reachable?

Reply:
Changed this method since this comment, but it is reachable now

If we are requesting multiple bodies, this will be reached right after we receive headers. At that point, we have not yet set a bodies request, so we need to set a bodies request with the hashes from the headers response